stdlib/LinearAlgebra/src/symmetric.jl

```diff
@@ -833,6 +833,13 @@ function eigen!(A::Hermitian{T,S}, B::Hermitian{T,S}; sortby::Union{Function,Not
     GeneralizedEigen(sorteig!(vals, vecs, sortby)...)
 end
 
+function eigen!(A::RealHermSymComplexHerm{T,S}, B::AbstractMatrix{T}; sortby::Union{Function,Nothing}=nothing) where {T<:Number,S<:StridedMatrix}
+    U = cholesky(B).U
+    vals, w = eigen!(U' \ A / U)
```

Comment:
```suggestion
    vals, w = eigen!(Hermitian(U' \ A / U))
```
Ideally, `U' \ A / U` would be computed with a specialized routine that exploited the structure of `A` and `U` but that doesn't have to be part of this PR.

Reply:
`cholesky(B).U` is of `UpperTriangular{T,S}` here, I thought this plus the type restriction of `A::RealHermSymComplexHerm{T,S} where {T<:Number,S<:StridedMatrix}` would already use the best possible algorithm (in this case, methods in [`triangular.jl`](https://github.com/JuliaLang/julia/blob/26a721b28ad56c006957ee1f2de083befcfe7904/stdlib/LinearAlgebra/src/triangular.jl#L1668) seem to be called). Do you mind suggesting a change? I would be happy to do it in this PR if that's not too complex.

Reply:
What I had in mind would be a function that applies `U` from the right and the left in one go. I.e. something like
![2021-01-18 15 50 31](https://user-images.githubusercontent.com/505001/104930222-27f86c00-59a5-11eb-938a-560abeacfafc.png)
where the last expression is a symmetric rank-one update and symmetric rank-two update.

Reply:
This is what I came up with:

```julia
"""
    f!(A, U)

Solves U' \ A / U for symmetric A and upper triangular U and stores the result in A.
"""
function f!(A, U)
    @inbounds for i in 1:size(A, 1)
        A[i, i] /= U[i, i]^2

        acoli = A[(i + 1):end, i]
        urowi = U[i, (i + 1):end]

        @views ldiv!(A[(i + 1):end, I],
                     LowerTriangular(U[(i + 1):end, (i + 1):end]'),
                     (@. acoli / U[i, i] - A[i, i] * urowi))
        A[i, (i + 1):end] = A[(i + 1):end, i]'

        @views @. A[(i + 1):end, (i + 1):end] +=
            -(urowi * acoli' + acoli * urowi') / U[i, i] + A[i, i] * (urowi * urowi')
    end
    return A
end
```

Unfortunately, this is no faster than `U' \ A / U`:

```julia
julia> A = rand(10, 10);

julia> A = A + A';

julia> B = A'A;

julia> U = cholesky(B).U;

julia> f!(copy(A), U) ≈ U' \ A / U
true

julia> using BenchmarkTools

julia> @benchmark f!($(copy(A)), U)
BenchmarkTools.Trial: 
  memory estimate:  9.38 KiB
  allocs estimate:  190
  --------------
  minimum time:     23.801 μs (0.00% GC)
  median time:      24.143 μs (0.00% GC)
  mean time:        26.156 μs (4.59% GC)
  maximum time:     1.120 ms (97.04% GC)
  --------------
  samples:          10000
  evals/sample:     6

julia> @benchmark U' \ A / U
BenchmarkTools.Trial: 
  memory estimate:  1.77 KiB
  allocs estimate:  3
  --------------
  minimum time:     1.355 μs (0.00% GC)
  median time:      1.393 μs (0.00% GC)
  mean time:        1.668 μs (3.42% GC)
  maximum time:     195.744 μs (97.91% GC)
  --------------
  samples:          10000
  evals/sample:     10
```

This is very much outside my expertise, so I very likely missed some obvious optimizations. Do you have any suggestions?

Reply:
For it to be competitive with the `U' \ A / U` version it is critical that the number of allocations is kept at a minimum so all the operations should be in-place. It can be convenient to use recursion as well so something like
```julia
function _UtiAsymUi!(uplo, A, U)
    n = size(A, 1)
    μ⁻¹ = U[1, 1]
    αμ2 = A[1,1]*μ⁻¹^2

    # Update (1,1) element
    A[1,1] = αμ2
    if n > 1
        if uplo === 'U'
            # update lower submatrix
            for j in 2:n
                aj = A[1, j]
                uj = U[1, j]
                for i in 1:j
                    ui = U[1, i]
                    A[i, j] = A[i, j] - ui'*μ⁻¹*aj - A[1, i]'*μ⁻¹*uj + ui'*αμ2*uj
                end
            end

            Unext = view(U, 2:n, 2:n)

            # Update vector
            for j in 2:n
                A[1, j] = A[i, j]*μ⁻¹ - U[1, j]*αμ2
                ldiv!(UpperTriangular(Unext)', view(A, 1, 2:n))
            end
        else
            ... something similar for 'L'
        end

        # Recurse
        _UtiAsymUi!(uplo, view(A, 2:n, 2:n), Unext)
    end

    return A
end

UtiAUi!(As::Symmetric, Utr::UpperTriangular) = _UtiAsymUi!(As.uplo, parent(As), parent(Ytr))
```
and a simplified version for `Diagonal` `U`.

Reply:
Perfect, thanks! Just rewrote and pushed the final code.